src/primitives/transaction.h

```diff
@@ -671,25 +656,16 @@ inline void UnserializeTransaction(TxType& tx, Stream& s) {
 
 template<typename Stream, typename TxType>
 inline void SerializeTransaction(const TxType& tx, Stream& s) {
-    const bool fAllowWitness = !(s.GetVersion() & SERIALIZE_TRANSACTION_NO_WITNESS);
-
     s << tx.nVersion;
-
     unsigned char flags = 0;
     // Consistency check
     assert(tx.wit.vtxoutwit.size() <= tx.vout.size());
-    if (fAllowWitness) {
-        /* Check whether witnesses need to be serialized. */
-        if (tx.HasWitness()) {
-            flags |= 1;
-        }
-    }
-    if (flags) {
-        /* Use extended format in case witnesses are to be serialized. */
-        std::vector<CTxIn> vinDummy;
-        s << vinDummy;
-        s << flags;
-    }
+
+    /* Check whether witnesses need to be serialized. */
+    if (tx.HasWitness())
```

Comment:
You need to check here whether `SERIALIZE_TRANSACTION_NO_WITNESS` was specified, and if so, leave flags at 0. Otherwise, your txids will commit to the witness data (making your transactions malleable again).